File: 477. Total Hamming Distance.cpp
class Solution {
public:
    // // brute force + hashmap: O(N2)time, O(1)space, TLE for large data
    // int totalHammingDistance(vector<int>& nums) {
    //     if(nums.empty()) return 0;
    //     int ret = 0;
    //     unordered_map<int, bitset<32>> map;
    //     for(int i = 0; i < nums.size(); i++)
    //         map[nums[i]] = bitset<32>(nums[i]);
    //     for(int i = 0; i < nums.size(); i++){
    //         bitset<32> a = map[nums[i]];
    //         for(int j = i + 1; j < nums.size(); j++){
    //             bitset<32> b = map[nums[j]];
    //             b ^= a;
    //             ret += b.count();
    //         }
    //     }
    //     return ret;
    // }

    // operate on 1 bit (rightmost) each time: M nums who have 1 as last bit and (n-M) who have 0 as
    // last bit will contribute M * (n-M) hamming distance. Iterate over 32 bits.
    // O(32N)time, O(1)space, 95.52%, ok
    int totalHammingDistance(vector<int>& nums) {
        if(nums.empty()) return 0;
        int ret = 0;
        for(int i = 0; i < 32; i++){
            int countBit = 0;
            for(int j = 0; j < nums.size(); j++)
                // check whether last bit is 1
                countBit += (nums[j] >> i) & 1;
            ret += countBit * (nums.size() - countBit);
        }
        return ret;
    }
};

File: 477. Total Hamming Distance.py
class Solution(object):
    
    # brute force: O(n2) time, O(1) space, TLE
    def totalHammingDistance(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        def hammingDistance(num1, num2):
            return bin(num1 ^ num2).count('1') 
        
        ret = 0
        for i in xrange(len(nums)):
            for j in xrange(i + 1, len(nums)):
                ret += hammingDistance(nums[i], nums[j])
        return ret 

    # math, count 1 in every bit: O(n) time, O(1) space
    def totalHammingDistance(self, nums):
        bits = {i:0 for i in range(32)}
        for num in nums:
            curr = bin(num)[2:]
            for i in xrange(len(curr)):
                bits[len(curr) - i - 1] += int(curr[i] == '1')
        ret = 0
        for count in bits.values():
            ret += count * (len(nums) - count)
        return ret 